### PR TITLE
Ouroboros nodeviewholder

### PR DESCRIPTION
**Pull request by i-Alex**
Changes in NodeViewHolder logic to support Ouroboros Praos with maximal similarity to scorex core for further core updates.

### CODE REVIEW

File: sdk/src/main/scala/com/horizen/SidechainNodeViewHolder.scala [hunk 1]
```diff
@@ -81,6 +93,134 @@ class SidechainNodeViewHolder(sidechainSettings: SidechainSettings,
...
+      if (state.version != branchingPoint) {
+        (
+          state.rollbackTo(idToVersion(branchingPoint)),
+          wallet.rollback(idToVersion(branchingPoint)),
+          trimChainSuffix(suffixApplied, branchingPoint)
+        )
+      } else (Success(state), Success(wallet), IndexedSeq())
+    } else (Success(state), Success(wallet), suffixApplied)
+
+    stateToApplyTry match {
```
Comment (oaleksandrov-ip) [inline sdk/src/main/scala/com/horizen/SidechainNodeViewHolder.scala]: (stateToApplyTry, walletToApplyTry) match {
      case (Success(stateToApply), Success(walletToApply)) =>
      case _ =>

File: sdk/src/main/scala/com/horizen/SidechainNodeViewHolder.scala [hunk 2]
```diff
@@ -81,6 +93,134 @@ class SidechainNodeViewHolder(sidechainSettings: SidechainSettings,
...
+  }
+
+  protected def requestDownloads(pi: ProgressInfo[SidechainBlock]): Unit = {
+    pi.toDownload.foreach { case (tid, id) =>
+      context.system.eventStream.publish(DownloadRequest(tid, id))
+    }
+  }
+
+  @tailrec
+  private def applyProgressInfo(history: HIS,
```
[the remark this answers is not in the document]
Comment (i-Alex) [reply sdk/src/main/scala/com/horizen/SidechainNodeViewHolder.scala]: actually now it is more like `updateStateAndWallet`

File: sdk/src/main/scala/com/horizen/SidechainNodeViewHolder.scala [hunk 3]
```diff
@@ -81,6 +93,134 @@ class SidechainNodeViewHolder(sidechainSettings: SidechainSettings,
   }
 
   override def receive: Receive = getCurrentSidechainNodeViewInfo orElse processLocallyGeneratedSecret orElse super.receive
+
+  override protected def pmodModify(pmod: SidechainBlock): Unit = {
```
[the remark this answers is not in the document]
Comment (i-Alex) [reply sdk/src/main/scala/com/horizen/SidechainNodeViewHolder.scala]: ok

File: sdk/src/main/scala/com/horizen/SidechainNodeViewHolder.scala [hunk 4]
```diff
@@ -81,6 +93,134 @@ class SidechainNodeViewHolder(sidechainSettings: SidechainSettings,
...
+        case Failure(e) =>
+          log.warn(s"Can`t apply persistent modifier (id: ${pmod.encodedId}, contents: $pmod) to history", e)
+          context.system.eventStream.publish(SyntacticallyFailedModification(pmod, e))
+      }
+    } else {
+      log.warn(s"Trying to apply modifier ${pmod.encodedId} that's already in history")
+    }
+  }
+
+  protected def requestDownloads(pi: ProgressInfo[SidechainBlock]): Unit = {
```
[the remark this answers is not in the document]
Comment (i-Alex) [reply sdk/src/main/scala/com/horizen/SidechainNodeViewHolder.scala]: ok

File: sdk/src/main/scala/com/horizen/SidechainNodeViewHolder.scala [hunk 5]
```diff
@@ -81,6 +93,134 @@ class SidechainNodeViewHolder(sidechainSettings: SidechainSettings,
...
+    if (idx == -1) IndexedSeq() else suffix.drop(idx)
+  }
+
+  // apply state and wallet with blocks one by one, if consensus epoch is going to be changed -> notify wallet and history.
+  protected def applyStateAndWallet(history: HIS,
+                           stateToApply: MS,
+                           walletToApply: VL,
+                           suffixTrimmed: IndexedSeq[SidechainBlock],
+                           progressInfo: ProgressInfo[SidechainBlock]): NodeUpdateInformation = {
+    val updateInfoSample = NodeUpdateInformation(history, stateToApply, walletToApply, None, None, suffixTrimmed)
```
[the remark this answers is not in the document]
Comment (i-Alex) [reply sdk/src/main/scala/com/horizen/SidechainNodeViewHolder.scala]: ok

File: sdk/src/main/scala/com/horizen/SidechainNodeViewHolder.scala (hunk 1 above)
Comment (i-Alex) [reply sdk/src/main/scala/com/horizen/SidechainNodeViewHolder.scala]: agree

File: sdk/src/main/scala/com/horizen/SidechainNodeViewHolder.scala [hunk 6]
```diff
@@ -81,6 +93,134 @@ class SidechainNodeViewHolder(sidechainSettings: SidechainSettings,
...
+    progressInfo.toApply.foldLeft(updateInfoSample) { case (updateInfo, modToApply) =>
+      if (updateInfo.failedMod.isEmpty) {
+        // Check if the next modifier will change Consensus Epoch, so notify History and Wallet with current info.
+        val (newHistory, newWallet) = if(updateInfo.state.isSwitchingConsensusEpoch(modToApply)) {
+          val (lastBlockInEpoch, stakeConsensusEpochInfo) = updateInfoSample.state.getCurrentStakeConsensusEpochInfo
+          (
+            updateInfo.history.applyStakeConsensusEpochInfo(lastBlockInEpoch, stakeConsensusEpochInfo),
```
[the remark this answers is not in the document]
Comment (i-Alex) [reply sdk/src/main/scala/com/horizen/SidechainNodeViewHolder.scala]: In my opinion, we don't need to do this, vals are already present in line 201, so context is pretty clear. It's a common behavior in our code (for example line 154)

File: sdk/src/main/scala/com/horizen/SidechainNodeViewHolder.scala [hunk 7]
```diff
@@ -81,6 +93,147 @@ class SidechainNodeViewHolder(sidechainSettings: SidechainSettings,
...
+          (historyAfterStakeConsensusApply, walletAfterStakeConsensusApply)
+        } else
+          (updateInfo.history, updateInfo.wallet)
+
+        updateInfo.state.applyModifier(modToApply) match {
+          case Success(stateAfterApply) =>
+            val walletAfterApply = newWallet.scanPersistent(modToApply)
+            val historyAfterApply = newHistory.reportModifierIsValid(modToApply)
+            context.system.eventStream.publish(SemanticallySuccessfulModifier(modToApply))
+            SidechainNodeUpdateInformation(historyAfterApply, stateAfterApply, walletAfterApply, None, None, updateInfo.suffix :+ modToApply)
```
Comment (oaleksandrov-ip) [inline sdk/src/main/scala/com/horizen/SidechainNodeViewHolder.scala]: look like we have here unnecessary copy here in "updateInfo.suffix :+ modToApply" we will do a deep copy of updateInfo.suffix plus modToApply. Shall we change suffix type from IndexedSeq to List?

File: sdk/src/main/scala/com/horizen/SidechainNodeViewHolder.scala [hunk 8]
```diff
@@ -81,6 +93,147 @@ class SidechainNodeViewHolder(sidechainSettings: SidechainSettings,
...
+            StakeConsensusEpochInfo(consensusEpochInfo.merkleTree.rootHash(), consensusEpochInfo.totalStake)
+          )
+          val walletAfterStakeConsensusApply = updateInfoSample.wallet.applyConsensusEpochInfo(consensusEpochInfo)
+          (historyAfterStakeConsensusApply, walletAfterStakeConsensusApply)
+        } else
+          (updateInfo.history, updateInfo.wallet)
+
+        updateInfo.state.applyModifier(modToApply) match {
+          case Success(stateAfterApply) =>
+            val walletAfterApply = newWallet.scanPersistent(modToApply)
```
Comment (albertog78) [inline sdk/src/main/scala/com/horizen/SidechainNodeViewHolder.scala]: move this after line 227

File: sdk/src/main/scala/com/horizen/SidechainNodeViewHolder.scala [hunk 9]
```diff
@@ -81,6 +93,147 @@ class SidechainNodeViewHolder(sidechainSettings: SidechainSettings,
...
+    progressInfo.toApply.foldLeft(updateInfoSample) { case (updateInfo, modToApply) =>
+      if (updateInfo.failedMod.isEmpty) {
+        // Check if the next modifier will change Consensus Epoch, so notify History and Wallet with current info.
+        val (newHistory, newWallet) = if(updateInfo.state.isSwitchingConsensusEpoch(modToApply)) {
+          val (lastBlockInEpoch, consensusEpochInfo) = updateInfoSample.state.getCurrentConsensusEpochInfo
+          val historyAfterStakeConsensusApply = updateInfo.history.applyStakeConsensusEpochInfo(
+            lastBlockInEpoch,
+            StakeConsensusEpochInfo(consensusEpochInfo.merkleTree.rootHash(), consensusEpochInfo.totalStake)
+          )
+          val walletAfterStakeConsensusApply = updateInfoSample.wallet.applyConsensusEpochInfo(consensusEpochInfo)
```
Comment (albertog78) [inline sdk/src/main/scala/com/horizen/SidechainNodeViewHolder.scala]: updateInfoSample -> updateInfo

**Pull request merged**